tracer/src/Datadog.Trace.Tools.Runner/Program.cs

```diff
@@ -57,159 +58,95 @@ internal static int Main(string[] args)
+            ciCommand.AddCommand(new RunCiCommand(applicationContext));
+            ciCommand.AddCommand(new CrankCommand());
 
-            config.AddCommand<AnalyzeInstrumentationErrorsCommand>("analyze-instrumentation")
-                  .WithDescription("Analyze instrumentation errors")
-                  .WithExample("analyze-instrumentation [--process-name dotnet]".Split(' '))
-                  .WithExample("analyze-instrumentation [--pid 12345]".Split(' '))
-                  .WithExample("analyze-instrumentation [--log-path \"C:\\ProgramData\\Datadog .NET Tracer\\logs\\\"]".Split(' '));
-
-            config.AddCommand<CoverageMergerCommand>("coverage-merge")
-                  .WithDescription("Merges all coverage json files into a single one.")
-                  .WithExample("coverage-merge c:\\coverage_folder\\ total-coverage.json".Split(' '))
-                  .IsHidden();
-        }
+            var checkCommand = new Command("check");
+            builder.Command.AddCommand(checkCommand);
 
-        private static int ExecuteLegacyCommandLine(string[] args, ApplicationContext applicationContext)
-        {
-            // Try executing the command with the legacy syntax
-            var app = new CommandApp<LegacyCommand>();
+            checkCommand.AddCommand(new CheckProcessCommand());
+            checkCommand.AddCommand(new CheckAgentCommand());
 
-            app.Configure(c =>
+            if (RuntimeInformation.IsOSPlatform(System.Runtime.InteropServices.OSPlatform.Windows))
             {
-                c.Settings.Registrar.RegisterInstance(applicationContext);
-                c.PropagateExceptions();
-            });
+                checkCommand.AddCommand(new CheckIisCommand());
+            }
 
-            return app.Run(args);
-        }
+            builder.Command.AddCommand(new RunCommand(applicationContext));
+            builder.Command.AddCommand(new AotCommand { IsHidden = true });
+            builder.Command.AddCommand(new AnalyzeInstrumentationErrorsCommand { IsHidden = true });
+            builder.Command.AddCommand(new CoverageMergerCommand { IsHidden = true });
 
-        // Extracted from Spectre.Console source code
-        // This is needed because we disable the default error handling to try the fallback legacy parser
-        private static List<IRenderable> GetRenderableErrorMessage(Exception ex, bool convert = true)
-        {
-            if (ex is CommandAppException renderable && renderable.Pretty != null)
-            {
-                return new List<IRenderable> { renderable.Pretty };
-            }
+            var parser = builder.Build();
 
-            if (convert)
+            var parseResult = parser.Parse(args);
+
+            if (parseResult.Errors.Count > 0)
             {
-                var converted = new List<IRenderable>
+                if (parseResult.Tokens.Count > 0 && parseResult.CommandResult.Command == builder.Command)
```

Comment:
I don't fully understand this fallback condition (I wouldn't have expected `parseResult.CommandResult.Command` to be non null if parsing failed?) but I take your (and the tests') word for it 😄

Reply:
I don't fully understand it either, but apparently, when the arguments are invalid, System.CommandLine resolves the root command.